### PR TITLE
Specify link attribute for Windows to be happy

### PR DESCRIPTION
**Pull request by seungha-yang**
import library needs to be explicitly linked

### CODE REVIEW

File: src/codegen/sys/lib_.rs [hunk 1]
```diff
@@ -23,6 +23,26 @@ pub fn generate(env: &Env) {
     save_to_file(&path, env.config.make_backup, |w| generate_lib(w, env));
 }
 
+fn get_library_name(lib: &String) -> String {
+    let mut s = String::from(lib);
+
+    if s.starts_with("lib") {
+        s = s.split_off("lib".len());
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:30]:
```suggestion
        s = &s[3..0];
```

Would also do :)

File: src/codegen/sys/lib_.rs [hunk 2]
```diff
@@ -23,6 +23,26 @@ pub fn generate(env: &Env) {
     save_to_file(&path, env.config.make_backup, |w| generate_lib(w, env));
 }
 
+fn get_library_name(lib: &String) -> String {
+    let mut s = String::from(lib);
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:27]: You don't need to create a `String` here, `&str` should be sufficient

File: src/codegen/sys/lib_.rs [hunk 3]
```diff
@@ -23,6 +23,26 @@ pub fn generate(env: &Env) {
...
+
+    if s.starts_with("lib") {
+        s = s.split_off("lib".len());
+    }
+
+    let offset = s.find(".so").unwrap_or(s.len());
+    s.truncate(offset);
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:34]: Or just

```suggestion
    &s[..offset]
```

:)

File: src/codegen/sys/lib_.rs [hunk 4]
```diff
@@ -23,6 +23,26 @@ pub fn generate(env: &Env) {
...
+    s
+}
+
+fn write_link_attr(w: &mut dyn Write, shared_libs: &Vec<String>) -> Result<()> {
+    for it in shared_libs.iter() {
+        writeln!(w, "#[link(name = \"{}\")]", get_library_name(it))?;
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:40]: I think this should also be used here: https://github.com/gtk-rs/gir/blob/a6737c72ae70ee7a17ae4ac7729905f9bb87e4e3/src/codegen/sys/cargo_toml.rs#L42

It's wrong what it does there currently.

File: src/codegen/sys/lib_.rs [hunk 5]
```diff
@@ -23,6 +23,26 @@ pub fn generate(env: &Env) {
...
+        s = s.split_off("lib".len());
+    }
+
+    let offset = s.find(".so").unwrap_or(s.len());
+    s.truncate(offset);
+    s
+}
+
+fn write_link_attr(w: &mut dyn Write, shared_libs: &Vec<String>) -> Result<()> {
+    for it in shared_libs.iter() {
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:39]:
```suggestion
    for it in &shared_libs {
```

**Comment by sdroege** [comment]
Apart from that various clippy warnings, you'll see them when clicking on the diff here or by running it locally :)

Otherwise looks good to me, thanks!

File: src/codegen/sys/lib_.rs (hunk 5 above)
Comment (seungha-yang) [reply src/codegen/sys/lib_.rs]: I guess you meant `for it in shared_libs` :) done

File: src/codegen/sys/lib_.rs (hunk 4 above)
Comment (seungha-yang) [reply src/codegen/sys/lib_.rs]: done

File: src/codegen/sys/lib_.rs [hunk 6]
```diff
@@ -23,6 +23,18 @@ pub fn generate(env: &Env) {
     save_to_file(&path, env.config.make_backup, |w| generate_lib(w, env));
 }
 
+fn write_link_attr(w: &mut dyn Write, shared_libs: &Vec<String>) -> Result<()> {
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:26]:
```suggestion
fn write_link_attr(w: &mut dyn Write, shared_libs: &[String]) -> Result<()> {
```

File: src/nameutil.rs [hunk 7]
```diff
@@ -125,6 +125,28 @@ pub fn lib_name_to_toml(name: &str) -> String {
...
+        s = &s[3..];
+    }
+
+    let offset = s.find(".so").unwrap_or(s.len());
+    s = &s[..offset];
+
+    s.to_string()
+}
+
+pub fn shared_libs_to_links(shared_libs: &Vec<String>) -> String {
```
Comment (sdroege) [inline src/nameutil.rs:141]:
```suggestion
pub fn shared_libs_to_links(shared_libs: &[String]) -> String {
```

File: src/nameutil.rs [hunk 8]
```diff
@@ -125,6 +125,28 @@ pub fn lib_name_to_toml(name: &str) -> String {
...
+    let mut links = vec![];
+    for it in shared_libs {
+        links.push(format!("\"{}\"", shared_lib_name_to_link_name(it)));
+    }
+
+    format!("[{}]", links.join(", "))
```
Comment (sdroege) [inline src/nameutil.rs:147]: Maybe only if `!links.is_empty()`?

File: src/nameutil.rs [hunk 9]
```diff
@@ -125,6 +125,28 @@ pub fn lib_name_to_toml(name: &str) -> String {
     name.to_string().replace("-", "_").replace(".", "_")
 }
 
+pub fn shared_lib_name_to_link_name(name: &str) -> String {
```
Comment (sdroege) [inline src/nameutil.rs:128]:
```suggestion
pub fn shared_lib_name_to_link_name(name: &str) -> &str {
```

Then you don't need the `to_string()` in the end

File: src/nameutil.rs (hunk 8 above)
Comment (seungha-yang) [reply src/nameutil.rs:147]: oh, empty list, good point :) but just curios it's possible case apart from programming error. (no native library is required for binding?)

Comment (sdroege) [reply src/nameutil.rs:147]: That's a good point. Should probably directly error out then instead

Comment (seungha-yang) [reply src/nameutil.rs]: now it will cause panic in that case

Comment (sdroege) [reply src/nameutil.rs]: Not nice to panic there but it seems like there's no other error reporting at all implemented right now...

File: src/nameutil.rs [hunk 10]
```diff
@@ -125,6 +125,36 @@ pub fn lib_name_to_toml(name: &str) -> String {
...
 }
 
+pub fn shared_lib_name_to_link_name(name: &str) -> &str {
+    let mut s = name;
+
+    if s.starts_with("lib") {
+        s = &s[3..];
+    }
+
+    let offset = s.find(".so").unwrap_or(s.len());
```
Comment (sdroege) [inline src/nameutil.rs:135]:
```suggestion
    let offset = s.find(".so").unwrap_or_else(|| s.len());
```

File: src/nameutil.rs (hunk 8 above)
Comment (seungha-yang) [reply src/nameutil.rs]: right, that's why I just gave up there 😅

File: src/nameutil.rs (hunk 10 above)
Comment (seungha-yang) [reply src/nameutil.rs]: done

**Review by sdroege — APPROVED** [review]

**Comment by seungha-yang** [comment]
`links = ["foo", "bar"]` or `['foo', 'bar']` doesn't work... not sure why. the last commit was updated so that picking up the first library from list

**Comment by sdroege** [comment]
> `links = ["foo", "bar"]` or `['foo', 'bar']` doesn't work... not sure why

Because cargo doesn't support it: https://github.com/rust-lang/cargo/issues/4533

File: src/nameutil.rs [hunk 11]
```diff
@@ -125,6 +125,35 @@ pub fn lib_name_to_toml(name: &str) -> String {
...
+        links.push(format!("\"{}\"", shared_lib_name_to_link_name(it)));
+    }
+
+    if links.is_empty() {
+        panic!("empty list of shared library");
+    }
+
+    // FIXME: links = ["foo", "bar"] or ['foo', 'bar'] doesn't seem to be working
+    // https://doc.rust-lang.org/cargo/reference/manifest.html#the-links-field
+    // is not mentioning about this though, probably cargo bug?
```
Comment (sdroege) [inline src/nameutil.rs:153]: https://github.com/rust-lang/cargo/issues/4533 says this is not supported

File: src/codegen/sys/cargo_toml.rs [hunk 12]
```diff
@@ -39,7 +39,9 @@ fn fill_empty(root: &mut Table, env: &Env, crate_name: &str) {
         set_string(
             package,
             "links",
-            nameutil::crate_name(&env.config.library_name),
+            // FIXME: Will pick the first library from list.
+            // links = ["foo", "bar"] doesn't work...
```
Comment (sdroege) [inline src/codegen/sys/cargo_toml.rs:43]: Remove the `FIXME` comment from here, that workaround is in the function itself and not here :)

File: src/nameutil.rs (hunk 11 above)
Comment (sdroege) [reply src/nameutil.rs:153]: Can you update the comment accordingly?

Looks good to me otherwise

**Review by sdroege — APPROVED** [review]

**Review by EPashkin — APPROVED** [review]
Checked build locally, it works

File: src/nameutil.rs [hunk 13]
```diff
@@ -125,6 +125,35 @@ pub fn lib_name_to_toml(name: &str) -> String {
...
+
+    if links.is_empty() {
+        panic!("empty list of shared library");
+    }
+
+    // https://github.com/rust-lang/cargo/issues/4533
+    // links = ["foo", "bar"] or ['foo', 'bar'] is not supported.
+    // Pick the first element
+    links[0].clone()
+}
```
Comment (GuillaumeGomez) [inline src/nameutil.rs:155]: Considering that `shared_lib_name_to_link_name` always return something, why not replacing this function with:

```suggestion
pub fn shared_libs_to_links(shared_libs: &[String]) -> String {
    // https://github.com/rust-lang/cargo/issues/4533
    // links = ["foo", "bar"] or ['foo', 'bar'] is not supported.
    // Pick the first element
    for it in shared_libs {
        return format!("\"{}\"", shared_lib_name_to_link_name(it));
    }
    panic!("empty list of shared library");
}
```

?

Comment (seungha-yang) [reply src/nameutil.rs]: that looks much better :) thanks for the suggestion.

Comment (sdroege) [reply src/nameutil.rs]: Causes a clippy warning though, rightfully:

```
 src/nameutil.rs#L145

this loop never actually loops
```

Comment (sdroege) [reply src/nameutil.rs]: So instead of the loop, just do `if !shared_libs.is_empty() { ... }`

Comment (seungha-yang) [reply src/nameutil.rs]: hmm why my cargo build/test build doesn't tell warning? :) btw updated

Comment (sdroege) [reply src/nameutil.rs]: Run `cargo clippy` to get extra warnings/suggestions :)

**Review by sdroege — APPROVED** [review]

**Comment by sdroege** [comment]
@GuillaumeGomez Let's get this in then?

File: src/nameutil.rs [hunk 14]
```diff
@@ -125,6 +125,30 @@ pub fn lib_name_to_toml(name: &str) -> String {
...
+    s
+}
+
+pub fn shared_libs_to_links(shared_libs: &[String]) -> String {
+    // https://github.com/rust-lang/cargo/issues/4533
+    // links = ["foo", "bar"] or ['foo', 'bar'] is not supported.
+    // Pick the first element
+    if !shared_libs.is_empty() {
+        return format!("\"{}\"", shared_lib_name_to_link_name(&shared_libs[0]));
+    }
```
Comment (GuillaumeGomez) [inline src/nameutil.rs:147]: Sorry, should have suggested:

```suggestion
if let Some(shared_lib) = shared_libs.iter().next() {
    return format!("\"{}\"", shared_lib_name_to_link_name(&shared_lib));
}
```

Comment (sdroege) [reply src/nameutil.rs:147]: Or even better: if let Some(item) = vec.get(0)

No need for an iterator.

Comment (GuillaumeGomez) [reply src/nameutil.rs:147]: True!

Comment (seungha-yang) [reply src/nameutil.rs]: thanks for the suggestions!

**Review by GuillaumeGomez — APPROVED** [review]
Looks good to me, thanks!

**Comment by sdroege** [comment]
@GuillaumeGomez You merge? Everybody approved it now :)

**Comment by GuillaumeGomez** [comment]
Normally it's up to @EPashkin but since they already approved, I guess it's fine.

**Pull request merged**